Fixed-length variants of MountainCar and CartPole

shwang opened the pull request:
Adds fixed-length variants of MountainCar-v0 and CartPole-v1.

~~MountainCar always has length 200 episodes and its rewards are 0 except upon reaching the flag, when its reward is 1.~~

~~CartPole always has length 500 episodes and its rewards are 0 except when falling over, when its reward is -1.~~

~~The reward modifications to the original environments were made using a new wrapper, `EpisodeEndRewardWrapper`.~~

I don't want to have to remember the format for the `gym.register` entry point or change these entries around if environments move between files, so I made a decorator (whose implementation turned out to be a bit nastier than I thought).

Let me know if this helper decorator is too involved.

edit:
`benchmark-environments/MountainCar-v0` always has length 200 episodes and its reward is -1 until it reaches the flag, after which point it enters an absorbing state with reward=0.

`benchmark-environments/CartPole-v0` always has length 500 episodes and its reward is 1 whenever it is in an okay state. An okay is any state which wouldn't have ended the episode in the original CartPole.

edit 2: Here are some Tensorboard plots of new environment RL/imitation learning perfomance.
[Notion document](https://www.notion.so/chaiberkeley/classical-control-benchmark-environments-52dc2b7d7b6e4b9cb7c95070b22ad7e4)

shwang commented:
I chose to make the fixed-length of Cartpole and MountainCar the same as the maximum episode length in the original envs. (No particular reason that I chose this).

For Cartpole this episode length has the nice property that average reward starts out in the range [-1, 0), and reaches 0 when Cartpole is expert.

shwang commented:
gym put these environments in the `gym.classic_control` module. Could rename to `classic_control` if this seems preferable

shwang commented:
Expert stats (generated via `python -m imitation.scripts.expert_demos`):

```
benchmark_environments/CartPole-v0

INFO - expert_demos - Result: {'n_traj': 56, 'return_min': 0.0, 'return_mean': 0.0, 'return_std': 0.0, 'return_max': 0.0, 'len_min': 500, 'len_mean': 500.0, 'len_std': 0.0, 'len_max': 500, 'monitor_return_min': 0.0, 'monitor_return_mean': 0.0, 'monitor_return_std': 0.0, 'monitor_return_max': 0.0
```

shwang commented:
Fixed-length cartpole + GAIL (takes roughly 4x more timesteps to achieve expert compared to variable-length environment):
![image](https://user-images.githubusercontent.com/1750835/74304637-71aaf200-4d12-11ea-8b34-abc250316757.png)

Fixed-length cartpole + AIRL (similar behavior to before, unstable. Succeeds in half of the seeds):
![image](https://user-images.githubusercontent.com/1750835/74305049-9eabd480-4d13-11ea-87a9-4a5da8a3ae93.png)

shwang commented:
Fixed-length Mount Car + GAIL (still broken) ([best of 3 video](https://youtu.be/vDc692VC2_M)):
![image](https://user-images.githubusercontent.com/1750835/74306005-2e528280-4d16-11ea-9423-c871354000fe.png)

Fixed-length Mount Car + AIRL (is fine):
![image](https://user-images.githubusercontent.com/1750835/74306656-fc422000-4d17-11ea-892a-a968e1233227.png)

AdamGleave commented on src/benchmark_environments/classic.py:
```diff
@@ -0,0 +1,32 @@
+"""Adaptation of classic Gym environments for IRL."""
```
I think I would prefer `classic_control`, in so far as we intend this to form part of the "fixed Gym" subset of tasks it's good to mirror their API. If we do keep it `classic`, then we'd probably also need to put in other "classic" environments like cliffwalk here.

AdamGleave replied:
Also nit: normally I see a newline after module docstring. (Although this does not actually seem to be in PEP8 or PEP257.)

AdamGleave commented on src/benchmark_environments/classic.py:
```diff
@@ -0,0 +1,32 @@
+"""Adaptation of classic Gym environments for IRL."""
+from gym.wrappers import TimeLimit
```
Google Python style (which we claim to largely be following in the README) is to always import modules, never classes/functions: see https://google.github.io/styleguide/pyguide.html#22-imports

So:
`from gym import wrappers`
and then later:
`wrappers.TimeLimit`.

This is one of the recommendations I'm less sure about, but I think it does generally help avoid name clash.

shwang replied:
This seems fine to me

AdamGleave commented on src/benchmark_environments/classic.py:
```diff
@@ -0,0 +1,32 @@
+
+
+@register("CartPole-v0")
+def cart_pole():
```
`cart_pole` and `mountain_car` are identical functions except for the environment names. Maybe make a functor? e.g.

```
def fixed_horizon(env_name):
    def make_env():
         env = util.make_env_no_wrappers(env_name)
         # ...
    return make_env

# ...

# If you never actually use register as a decorator, maybe change it to be register(our_name, env_thunk_fn) rather than register(our_name)(env_thunk_fn)
register("CartPole-v0")(fixed_horizon("CartPole-v1")
```

AdamGleave replied:
Ah I see they differ in sign of the final reward, but that seems easy to add as a parameter.

shwang replied:
I don't think the functor is worth making until there is more duplication (IE, at least three functions of this type or longer functions duplicated).

I removed the "automatically detect the time limit of cartpole/mountaincar" logic so that the code can match up with the docstring, and this reduces the number of lines duplicated.

AdamGleave commented on src/benchmark_environments/mujoco.py:
```diff
@@ -11,6 +11,8 @@
     walker2d_v3,
 )
 
+import benchmark_environments.util as util
```
```suggestion
from benchmark_environments import util
```
feels slightly more idiomatic to me

AdamGleave commented on src/benchmark_environments/mujoco.py:
```diff
@@ -72,5 +74,5 @@ def _get_gym_v3_max_episode_steps(env_base_):
         id=f"benchmark_environments/{env_base}-v0",
         entry_point=f"benchmark_environments.mujoco:{env_base}Env",
-        max_episode_steps=_get_gym_v3_max_episode_steps(env_base),
+        max_episode_steps=util.get_gym_max_episode_steps(f"{env_base}-v3"),
```
Can you delete `_get_gym_v3_max_episode_steps` from `mujoco.py`?

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -1,4 +1,7 @@
 """Miscellaneous utilities."""
```
Nit: newline after module docstring.

(I could be persuaded away from this given it doesn't seem to be PEP8/PEP257, but then we should change the other files for consistency.)

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +20,107 @@ def step(self, action):
+class EpisodeEndRewardWrapper(gym.Wrapper):
+    """Replaces all rewards, with all rewards becoming zero except episode end.
+
+    Useful for converting living rewards into equivalent episode termination
```
Can you make explicit the episode terminates when some special condition is reached, and not (in its intended use case) when it timesout?

If episode termination is *always* reached, then in the undiscounted case all policies are optimal, and even in the discounted case the incentives are quite weak.

shwang replied:
I removed this wrapper (because creates non-reward equivalent variants of gym envs) so this comment is now n/a

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +20,107 @@ def step(self, action):
+            after seeding.
+        max_episode_steps (Optional[int]): The maximum number of steps that an episode
+            can consist of.
+        kwargs (dict): Kwargs to pass into `func`.
```
Should this be `register_kwargs`?

shwang replied:
No, the register_kwargs argument to `gym.register` in fact accepts the key `kwargs`. I'll make a note of this.

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +20,107 @@ def step(self, action):
+        module_name: The name of the module where the decorator will be applied.
+            Usually the user will want to use `__name__`.
+        id_prefix: The Gym-registered `id` is `id_prefix + id_base`.
+        reward_threshold (Optional[int]): The reward threshold before the task is
```
The next three arguments don't appear in the function signature, I assume they're passed-through to `gym.register`?

shwang replied:
That's right, they are default arguments to `EnvSpec`. I'll make a note of this before the arguments section.

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +20,107 @@ def step(self, action):
+    return func
+
+
+def curried_gym_register_as_decorator(module_name: str,) -> Callable[[str], Callable]:
```
```suggestion
def curried_gym_register_as_decorator(module_name: str) -> Callable[[str], Callable]:
```
Not normally trailing commas for one-line function signature

shwang replied:
thanks for fixing the linter too

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -1,4 +1,7 @@
 """Miscellaneous utilities."""
+from functools import partial
```
```suggestion
import functools
```
`functools.partial` is not that much longer, and it'll be much clearer what's going on.

AdamGleave commented on tests/test_envs.py:
```diff
@@ -3,6 +3,7 @@
 import gym
 import pytest
 
+import benchmark_environments.classic  # noqa: F401
```
Consider clarifying why we're importing this:
```
# Unused (noqa: F401) imports required for env registration
import benchmark_environments.classic  # noqa: F401
import benchmark_environments.mujoco  # noqa: F401
```

AdamGleave commented on tests/test_wrappers.py:
```diff
@@ -1,7 +1,8 @@
 
 from benchmark_environments.testing.envs import CountingEnv
-from benchmark_environments.util import AutoResetWrapper
+from benchmark_environments.util import AutoResetWrapper, EpisodeEndRewardWrapper
```
Reiterating https://google.github.io/styleguide/pyguide.html#22-imports comment (although in this file see we're already importing classes)

shwang replied:
Fixed in both cases

AdamGleave commented:
My analysis was assuming at the normal episode termination you transition into an absorbing state. I've realised the current implementation actually reset (silently). This complicates the analysis considerably.

Assuming undiscounted case, for CartPole I think the expected return in the original Gym was the mean time to failure (MTTF). If we have equal probability of failing at any timestep (clearly not true, I think impossible to fail at the first timestep), then the expected return in the new environment I think is just -1/MTTF.

These feel non-equivalent. In these dynamics probably will induce similar optimal policies. But suppose you could flip a coin at the start and have 50% chance of never failing, and 50% chance of dying immediately. In original Gym (if we removed the episode time limit) flipping the coin is great (infinite expected returns), but in our modified environment you wouldn't want to do that (-1/2 return).

We should think more about whether we want to mantain Gym compatibility. If not I think we can come up with a more sensible reward function than either of these. If we do then we'll probably need to do something with absorbing states.

shwang commented:
Thanks for the review, plan on making these changes tonight or the beginning of tomorrow.

shwang commented:
IIRC we thought on Thurs night last week that it would be fine to leave these environments in the repository on the understanding that they are preliminary, in-flux environments used for `imitation` training, and we will document that they are very much not affine-transform-equivalent to the original gym environments.

AdamGleave commented:
> IIRC we thought on Thurs night last week that it would be fine to leave these environments in the repository on the understanding that they are preliminary, in-flux environments used for `imitation` training, and we will document that they are very much not affine-transform-equivalent to the original gym environments.

I'm OK with this, although for the purpose of `imitation` testing I still wonder if these environments are really what we want. What do you make of @qxcv's suggestion to make each episode fixed horizon, so on `CartPole` it does not terminate if the pole falls (you just stop getting reward) and in `MountainCar` the train stops once it reaches the flag?

shwang commented:
> > IIRC we thought on Thurs night last week that it would be fine to leave these environments in the repository on the understanding that they are preliminary, in-flux environments used for `imitation` training, and we will document that they are very much not affine-transform-equivalent to the original gym environments.
> 
> I'm OK with this, although for the purpose of `imitation` testing I still wonder if these environments are really what we want. What do you make of @qxcv's suggestion to make each episode fixed horizon, so on `CartPole` it does not terminate if the pole falls (you just stop getting reward) and in `MountainCar` the train stops once it reaches the flag?

That seems fine to me too. (bonus for not biasing expert demonstrations towards early states). I'lll work on making this change after I have some heatmaps for the current iteration of the fixed length environments.

shwang commented:
> on `CartPole` it does not terminate if the pole falls (you just stop getting reward) and in `MountainCar` the train stops once it reaches the flag?

I've implemented these environments with new wrappers (and removed the old wrapper that made the non-reward-equivalent fixed-length wrappers.)

Not sure why black is failing here, it seems fine on my machine.

AdamGleave commented on src/benchmark_environments/classic_control.py:
```diff
@@ -0,0 +1,37 @@
+    In the event of early episode completion (IE, the car reaches the
+    goal), we enter an absorbing state that repeats the final observation
+    and returns reward 0.
+    Done is always returned on timestep 200 only.
```
```suggestion

    Done is always returned on timestep 200 only.
```

AdamGleave commented on src/benchmark_environments/classic_control.py:
```diff
@@ -0,0 +1,37 @@
+def cart_pole():
+    """Fixed-length variant of CartPole-v1.
+
+    If the agent fails (IE pole falls over or CartPole exits the screen),
```
```suggestion
    If the agent fails (i.e. pole falls over or CartPole exits the screen),
```

AdamGleave commented on src/benchmark_environments/classic_control.py:
```diff
@@ -0,0 +1,37 @@
+def mountain_car():
+    """Fixed-length variant of MountainCar-v0.
+
+    In the event of early episode completion (IE, the car reaches the
```
```suggestion
    In the event of early episode completion (i.e., the car reaches the
```

AdamGleave commented on src/benchmark_environments/classic_control.py:
```diff
@@ -0,0 +1,37 @@
+    """Fixed-length variant of CartPole-v1.
+
+    If the agent fails (IE pole falls over or CartPole exits the screen),
+    all rewards become 0 for the duration of the episode.
```
This does mean the reward is non-Markovian which is a little strange (although plausibly fairly common in practice). Possible alternatives:
  1. Just let the reward be positive if it manages to get the pole upright again (it probably won't!)
  2. Use the `AbsorbAfterDoneWrapper`.

@qxcv do you have thoughts on this?

qxcv replied:
I doubt it will make much difference to RL. That said, I'd prefer alternative (1) or (2) because those align with the original motivations of the benchmark reimp project (e.g. removing hidden state from reward functions, etc.).

AdamGleave replied:
OK. In that case I think I'd lean towards (1) since this seems more physically realistic, but ok with (2) as well if it's much easier to implement.

shwang replied:
(1) seems fine to me. I have a clarifying question about what counts as upright which I'm asking on Slack

shwang replied:
It looks like the cleanest way to implement (1) is to subclass Gym's CartPoleEnv and override `step()`.

AdamGleave replied:
I think rotating 360 degrees so it ends up vertically upwards again should still count as upright. This is analogous to a robotic "swing-up" task.

I can see a case for disallowing this, but I think we'd then need to augment the state space to indicate it if it ever fell over, otherwise the reward would be non-Markovian.

Hope you're feeling better!

shwang replied:
Two weird things I wanted to bring up:
(a) Under the original implementation (with termination condition removed), the meaning of the CW and CCW actions becomes reversed when the pole is under the cart. Is this worth fixing? (Don't expect RL algorithms to be hurt by this, but could be confusing. Fixing probably just means adding another if statement.)

(b) I'd expect that if I removed all force except gravity that the pole would swing back and forth in an arc, [but what actually happens](https://www.youtube.com/watch?v=MluW34l_g5I&feature=youtu.be) is that the pole completes the entire circle, and in fact starts accelerating with each completion. Even if I force initial theta velocity to be 0 and reduce tau by 10x ("seconds between state updates"). Probably due to the eulers method approximation. I'm not expecting this to affect RL.

shwang replied:
> I can see a case for disallowing this, but I think we'd then need to augment the state space to indicate it if it ever fell over, otherwise the reward would be non-Markovian.

trivia: Actually, in the original implementation but with terminal state removed, we are in a distinct state when we swing 360 degrees around. The pole theta appears as something like 2*pi rather than 0 when it is at the top because the angle is never normalized.

AdamGleave replied:
> Two weird things I wanted to bring up:
> (a) Under the original implementation (with termination condition removed), the meaning of the CW and CCW actions becomes reversed when the pole is under the cart. Is this worth fixing? (Don't expect RL algorithms to be hurt by this, but could be confusing. Fixing probably just means adding another if statement.)

As in the CW and CCW actions exert torque in different directions depending on the y-coordinate? I'd be inclined to fix this so it's more physically realistic. Agree it probably won't affect RL (optimal policy is to never fall over, after all) though.

> (b) I'd expect that if I removed all force except gravity that the pole would swing back and forth in an arc, [but what actually happens](https://www.youtube.com/watch?v=MluW34l_g5I&feature=youtu.be) is that the pole completes the entire circle, and in fact starts accelerating with each completion. Even if I force initial theta velocity to be 0 and reduce tau by 10x ("seconds between state updates"). Probably due to the eulers method approximation. I'm not expecting this to affect RL.

Are you sure you haven't just invented a perpetual motion machine? ;) Yeah this doesn't seem worth fixing.

shwang replied:
> (a) Under the original implementation (with termination condition removed), the meaning of the CW and CCW actions becomes reversed when the pole is under the cart. Is this worth fixing? (Don't expect RL algorithms to be hurt by this, but could be confusing. Fixing probably just means adding another if statement.)

This was my misunderstanding -- there's aren't CW/CCW actions on the pole. Rather, there are left/right actions on the cart.

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +22,183 @@ def step(self, action):
+
+
+class AbsorbAfterDoneWrapper(gym.Wrapper):
+    """Goes to absorb state after done=True. Always returns done=False.
```
```suggestion
    """Transition into absorbing state instead of episode termination.
    
    When the environment being wrapped returns `done=True`, we return an absorbing observation.
    This wrapper always returns `done=False`.
```

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +22,183 @@ def step(self, action):
+                # starting on the next call to `step()`.
+                self.at_absorb_state = True
+
+                if self.absorb_obs_default is None:
```
Inclined to make this a one-liner, although maybe it's too long...

```suggestion
                self.absorb_obs_this_episode = inner_obs if self.absorb_obs_default is None else self.absorb_obs_default
```

shwang replied:
It's too long (length 120). Tried breaking it up with "\", but I mildly prefer the current formatting:

![image](https://user-images.githubusercontent.com/1750835/76057234-bac71e00-5f2d-11ea-8a6e-530be6ac4bca.png)

shwang replied:
black doesn't accept breaking up with `\`. Automatically formats to
![image](https://user-images.githubusercontent.com/1750835/76122276-89406800-5faa-11ea-965a-75ce3882ba28.png)

I'm sticking to the original format.

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +22,183 @@ def step(self, action):
+            assert self.absorb_reward is not None
+            obs = self.absorb_obs_this_episode
+            rew = self.absorb_reward
+            info = {}  # XXX: Also allow custom info for absorb state.
```
No info in absorbing state doesn't seem like a big deal.

If you want to leave a comment please make it `TODO(shwang): ...`

Generally todo's should always have some source of info for context. Ideally a GitHub issue, but a username better than nothing.

shwang replied:
Removed the todo comment and added to docstring that info dict is always empty.

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +22,183 @@ def step(self, action):
+) -> Callable[..., gym.Env]:
+    """Decorator variant of `gym.register` with autogenerated `id` and `entry_point`.
+
+    (`reward_threshold`, `nondeterministic`, `max_episode_steps`, and `kwargs` are
```
I still find this comment a bit confusing (documenting another functions arguments). How about:

```python
`**register_kwargs` is passed through directly to `gym.register`. This function
takes optional arguments `reward_threshold`, `nondeterministic`, `max_episode_steps`,
and `kwargs`, documented below.
```

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +22,183 @@ def step(self, action):
+def curried_gym_register_as_decorator(module_name: str) -> Callable[[str], Callable]:
+    """Two-staged curry that builds a easy-to-use `gym.register` decorator.
+
+    See `benchmark_environments.classic` for example usage.
```
```suggestion
    See `benchmark_environments.classic_control` for example usage.
```

AdamGleave commented on tests/test_wrappers.py:
```diff
@@ -29,3 +31,73 @@ def test_auto_reset_wrapper(episode_length=3, n_steps=100, n_manual_reset=2):
+def test_absorb_repeat_custom_state(
+    absorb_reward=-4, absorb_obs=-3.0, episode_length=6, n_steps=100, n_manual_reset=3,
+):
+    """Check that AbsorbAfterDoneWrapper returns custom state."""
```
```suggestion
    """Check that AbsorbAfterDoneWrapper returns custom state and reward."""
```

AdamGleave commented on tests/test_wrappers.py:
```diff
@@ -29,3 +31,73 @@ def test_auto_reset_wrapper(episode_length=3, n_steps=100, n_manual_reset=2):
+        env, absorb_reward=absorb_reward, absorb_obs=absorb_obs,
+    )
+
+    obs_list = []
```
this looks to be left over from debugging

shwang replied:
Thanks for catching this

AdamGleave commented on tests/test_wrappers.py:
```diff
@@ -29,3 +31,73 @@ def test_auto_reset_wrapper(episode_length=3, n_steps=100, n_manual_reset=2):
+        for t in range(1, n_steps + 1):
+            act = env.action_space.sample()
+            obs, rew, done, _ = env.step(act)
+            obs_list.append(obs)  # DEBUG
```
remove?

shwang commented:
Here are some plots of RL/imitation performance. (Link also added to the PR description)

https://www.notion.so/chaiberkeley/classical-control-benchmark-environments-52dc2b7d7b6e4b9cb7c95070b22ad7e4

AdamGleave commented on src/benchmark_environments/classic_control.py:
```diff
@@ -0,0 +1,64 @@
+        x, _, theta, _ = self.state
+
+        # Normalize theta to [-pi, pi] range.
+        theta += np.pi
```
A one-liner might be easier to read?
```suggestion
        theta = (theta + np.pi) % (2 * np.pi) - np.pi
```

AdamGleave commented on src/benchmark_environments/classic_control.py:
```diff
@@ -0,0 +1,64 @@
+        theta -= np.pi
+
+        state_fail = bool(
+            x < -self.x_threshold
```
```suggestion
            abs(x) > self.x_threshold
            or abs(theta) > self.theta_threshold_radians
```

shwang replied:
I also negated the state to for test `state_ok` instead of `state_fail`

AdamGleave commented on src/benchmark_environments/util.py:
```diff
@@ -17,3 +22,154 @@ def step(self, action):
+def make_env_no_wrappers(env_name: str, **kwargs) -> gym.Env:
+    """Gym sometimes wraps envs in TimeLimit before returning from gym.make().
+
+    This helper method builds directly from step to avoid this wrapper.
```
```suggestion
    This helper method builds directly from spec to avoid this wrapper.
```

AdamGleave commented on src/benchmark_environments/testing/envs.py:
```diff
@@ -211,11 +211,9 @@ def step(self, action):
         """Step method for CountingEnv."""
         if self.timestep is None:
             raise RuntimeError("Need to reset before first step().")
-        if self.timestep > self.episode_length:
```
Why are we removing this? I don't think we call `step()` on `CountingEnv` after `done` is True?

shwang replied:
Removed this error so I can testing `AbsorbAfterDone` wrapper, which needs environment to continue progressing after the first `done=True`.

shwang replied:
Hmmmm that sounds right actually

shwang replied:
Fixed, thanks for catching this

shwang commented:
I've added the AIRL/GAIL plots for the cartpole benchmark environment to the notion link I posted before.

AIRL looks great, taking about 3x timesteps compared to RL to reach expert. GAIL performance collapses under our default hyperparams, although in one seed out of 8, it reached returns of near 500 before collapsing.

AdamGleave commented:
> AIRL looks great, taking about 3x timesteps compared to RL to reach expert. GAIL performance collapses under our default hyperparams, although in one seed out of 8, it reached returns of near 500 before collapsing.

So the tl;dr is that AIRL works well in both environments and GAIL struggles in both? This is at least more consistent than the pattern before.

AdamGleave commented:
Strangely the `patch/tests` check is failing despite `project/tests` having 100% coverage. I can't see what's triggering this so it's probably fine to ignore, codecov has been quite flaky lately.

shwang commented:
@AdamGleave 
It turns out that this link wasn't actually relevant in this case, but wanted to resurface this because it was hard to look for. IDK if you saw this comment but shortly after we moved to CircleCI on imitation I noticed that shard 3 out of 3 was ignored in coverage counts: https://github.com/HumanCompatibleAI/imitation/pull/142#issuecomment-567745513

AdamGleave commented:
Interesting. I've observed some times that `codecov` fails to update after additional shards report but it's seemed non-deterministic. Do you know if this was happening consistently?

We may just want to switch to another platform, although the alternatives also seem to have their share of problems.

shwang commented:
Unlikely that it's happening consistently, because codecov seems to fail only intermittently on imitation.
